### PR TITLE
Enable the ability to add an IO tile with registers with no configuration

### PR DESCRIPTION
**Pull request by jack-melchert**
Alex asked for an IO tile with registers that didn't have any configuration, so these changes are necessary to get that working.

### CODE REVIEW

File: canal/global_signal.py [hunk 1]
```diff
@@ -117,6 +116,7 @@ def apply_global_meso_wiring(interconnect: Interconnect, io_sides: IOSide = IOSi
         # along with OR gate to reduce input read_data with
         # that tile's read_data
         # ports_in keep track of new ports created by or_reduction
+        column = [entry for entry in column if "config" in entry.ports]
```
Comment (Kuree) [inline canal/global_signal.py:119]: Is there a reason to reorganize this statement?

File: canal/circuit.py [hunk 2]
```diff
@@ -938,18 +940,27 @@ def __add_reset(self):
...
         # or not
         for feature in self.features():
             if "config" in feature.ports:
                 return True
-            else:
-                # if the feature doesn't have config port, it shouldn't have
-                # reset either, although the other way around may be true
-                # that is, a feature may have some internal states that need
-                # to reset, but not necessarily has config port
-                assert "reset" not in feature.ports
```
Comment (Kuree) [inline canal/circuit.py:952]: Why is this removed?

File: canal/circuit.py [hunk 3]
```diff
@@ -938,18 +940,27 @@ def __add_reset(self):
         for reset_port in reset_ports:
             self.wire(self.ports.reset, reset_port)
 
+    def __add_clk(self):
+        # automatically add clk signal and connect it to the features if the
+        # feature supports it
+        clk_ports = []
```
Comment (Kuree) [inline canal/circuit.py]: Set?

File: canal/circuit.py [hunk 4]
```diff
@@ -848,7 +848,9 @@ def __init__(self, tiles: Dict[int, Tile],
         self.__add_tile_id()
         # add ports
         self.add_ports(stall=magma.In(magma.Bits[stall_signal_width]),
-                       reset=magma.In(magma.AsyncReset))
+                       reset=magma.In(magma.AsyncReset),
+                       clk=magma.In(magma.Clock),
```
Comment (Kuree) [inline canal/circuit.py:852]: If the tile doesn't have clock, don't you need to remove the port?

**Review by Kuree — CHANGES_REQUESTED** [review]

File: canal/global_signal.py (hunk 1 above)
Comment (jack-melchert) [reply canal/global_signal.py:119]: A register tile with a pipelining register but no config has a clk and reset port but no config port. This pass skips the IO tile clk and reset signals if the statement is left where it was originally.

File: canal/circuit.py (hunk 2 above)
Comment (jack-melchert) [reply canal/circuit.py:952]: New IO tile has a clk and reset signal but no config.

File: canal/circuit.py (hunk 4 above)
Comment (jack-melchert) [reply canal/circuit.py:852]: all tiles have stall, reset, and clk ports right?

Comment (Kuree) [reply canal/circuit.py:852]: Not necessarily. In Jade we have empty tiles that doesn't contain anything, since we have IOs on four sides. I think there is code somewhere that pops these ports. Just to make sure that these ports are popped correctly.

Comment (jack-melchert) [reply canal/circuit.py:852]: Sure, that code is in the garnet repo I believe. Is that the appropriate place for it?

**Comment by jack-melchert** [comment]
Is this ready to merge?

**Review by Kuree — APPROVED** [review]
